src/conventional/sonde_bufr2ioda.py

```diff
@@ -430,181 +415,309 @@ def read_bufr_message(f, count, start_pos, data):
+                temp_data[variable] = None
+
+    if not repfacs:
+        for variable in raw_obsvars:
+            if repfactors[variable]:
+                repfacs = repfactors[variable]
+                break
+
+    # From repfacs, make begin/end indicies for single or multiple soundings.
+    nbeg = []
+    nend = []
+    if repfacs:
+        if nsubsets > 1:
+            if nsubsets != len(repfacs):
+                logging.warning(f"Nonsense: number of subsets, {nsubsets} is not equal to "
+                                f"the length of repfacs vector, {len(repfacs)}")
+            nend = np.cumsum(repfacs)
+            nbeg = np.insert(nend[:-1], 0, 0)
+        else:
+            nbeg.append(0)
+            nend.append(repfacs[0]-1)
+    else:
+        nbeg.append(0)
+        nend.append(int(1E6))
 
     # Be done with this BUFR message.
     ecc.codes_release(bufr)
 
-    # Need to transform some variables to others (wind speed/direction to components for example).
-    uwnd = np.full(target_number, float_missing_value)
-    vwnd = np.full(target_number, float_missing_value)
-    for n, wdir in enumerate(vals['windDirection']):
-        wspd = vals['windSpeed'][n]
-        if (wdir >= 0 and wdir <= 360 and wspd >= 0 and wspd < 300):
-            uwnd[n], vwnd[n] = met_utils.dir_speed_2_uv(wdir, wspd)
-
-    spfh = np.full(target_number, float_missing_value)
-    for n, dewpoint in enumerate(vals['dewpointTemperature']):
-        pres = meta_data['pressure'][n]
-        if (dewpoint > 50 and dewpoint < 325 and pres > 100 and pres < 109900):
-            spfh[n] = met_utils.specific_humidity(dewpoint, pres)
-
-    # Very odd, sometimes the first level of data has some variables set to zero. Reset to missing.
-    if (meta_data['geopotentialHeight'][0] == 0 or meta_data['pressure'][0] == 0):
-        meta_data['geopotentialHeight'][0] = float_missing_value
-        meta_data['pressure'][0] = float_missing_value
-    mask_T = np.logical_or(vals['airTemperature'] < 50, vals['airTemperature'] > 335)
-    vals['airTemperature'][mask_T] = float_missing_value
-
-    # Move everything into the final data dictionary, including metadata.
-    data['windEastward'] = np.append(data['windEastward'], uwnd)
-    data['windNorthward'] = np.append(data['windNorthward'], vwnd)
-    data['specificHumidity'] = np.append(data['specificHumidity'], spfh)
-    data['airTemperature'] = np.append(data['airTemperature'], vals['airTemperature'])
-    for key in meta_keys:
-        data[key] = np.append(data[key], meta_data[key])
+    # Loop over each pair of beginning and ending indices and transfer data
+    obnum = 0
+    for b, e in tuple(zip(nbeg, nend)):
+        if b < 0 or e < 0 or e <= b:
+            logging.warning(f"Skipping nonsense BUFR msg with a negative index [{b},{e}]")
+            return data, count, start_pos
+        logging.debug(f"Within BUFR msg, processing ob {obnum+1} with bounds: [{b},{e-1}]")
+        if e < 999999:
```

Comment:
Line number 552 sets the ending element to 1E6 as a default behavior.  So this is the "catch" to when that value is used.  This is so annoying in BUFR because we can never ensure how many observations might be sent in a message.  So this is another catch for bad things.  And Python is utterly annoying because when I did not use ```int(1E6) in line 552, python somehow takes it to mean a floating point not an integer.  Weakly-typed languages have serious annoyances.

Reply:
I added a couple sentences in the comment.